README.md
surviplot
========

An R package for drawing survival curves with the number-at-risk displayed underneath.

The primary function "surviplot" draws a survival curve with the number at risk 
displayed underneath. Optionally, the hazard ratio, CI, and logrank P value are 
displayed if there are exactly two strata. 

You can see a couple of examples here:
http://www.cbs.dtu.dk/~eklund/surviplot/



History: surviplot vs. survplot
-------------------------------

I originally called this package "survplot" and made it available on my website
around 2009.  Soon after, I discovered the existence of a function called "survplot"
in the [rms package](https://cran.r-project.org/package=rms) with
somewhat similar functionality.  I was heartbroken.  Worse, there seemed to be people
out there who were getting the two packages confused. 

After seven years of anguish I finally arrived at a satisfactory new name for my
package: **surviplot**.  

Version 1.0.0 of surviplot is essentially the same as version 0.0.7 of
survplot, except for the "survplot" -> "surviplot" thing.

For former users of "survplot", you just need to replace

    ## old
    library(survplot)
    survplot(...)

with 

    ## new
    library(surviplot)
    surviplot(...)



Installation
------------


You can install the latest version from GitHub like this:

    ## install.packages("devtools")
	library("devtools")
	install_github("aroneklund/surviplot")


Related works
-------------

There are other R packages with similar functionality:
* [**rms**](https://cran.r-project.org/package=rms) also draws
survival curves with number-at-risk.
* [**survminer**](https://cran.r-project.org/package=survminer)
draws survival curves with number-at-risk in the ggplot2 framework.




NEWS
surviplot NEWS


Changes in version 1.1.0 (2017-04-05)

- Modified "cutn" such that it doesn't choke on challenging data (or at least chokes less often).


Changes in version 1.0.0 (2016-08-08)

- This package was formerly called "survplot" (until version 0.0.7) and is now called "surviplot". This was done to avoid confusion with the function "survplot" in the "rms" package.
- Similarly, the function "survplot" is now called "surviplot".


Changes in version 0.0.7 (2012-04-04)

- Unstratified data should now work; see new examples in documentation.
- "cutn" now accepts NAs without complaining, and now does not have an "na.rm" argument.
- New function "addNrisk" for adding number-at-risk to an existing plot.


Changes in version 0.0.6 (2012-02-20)

- Added a NAMESPACE file.


Changes in version 0.0.5 (2011-07-06)

- In "survplot", supplying a list to legend.pos would generate a warning.
- "survplot" now returns HR and P value (as text), invisibly (if there are two groups).


Changes in version 0.0.4 (2010-10-20)

- Updated keywords in function docs
- Modernized this NEWS file (was ChangeLog)
- Changed arg in "survplot": "show.n.risk" is now "show.nrisk".
- Added "color.nrisk" arg in "survplot"
- Moved "Number at risk" label to the left
- Moved number at risk section up one line to take up less space.
- In DESCRIPTION, added link to the survplot web page.


Changes in version 0.0.3 (2009-06-24)

- Added "invert" function.
- "plotsurv": Check length of 'snames'.
- "plotsurv": Automatically trim redundant prefix of 'snames'.
- "plotsurv": Add 'stitle' and 'lwd' arguments.


Changes in version 0.0.2 (2009-04-29)

- "nrisk" was crashing in R-2.9.0 because "survfit.object" changed. 
- "plotsurv" was crashing in R-2.9.0 because "plot.survfit" can no longer make a legend.
- Added "lty" and "legend.pos" arguments to "plotsurv".


Changes in version 0.0.1 (2009-02-20)

- First release.  The functions were all taken from the (not publicly available) "aronmisc" package.



R/survplot.R
# surviplot.R
#
# Aron Eklund   eklund@cbs.dtu.dk
#
# formerly known as "survplot.R" until 2016-08-08
#
#


nrisk <- function(x, times = pretty(x$time)) { 
    stopifnot(class(x) == 'survfit')
    if('strata' %in% names(x)) {
      ns <- length(x$strata)
      idx <- rep.int(1:ns, x$strata)
      str.n.risk <- split(x$n.risk, idx)
      str.times <- split(x$time, idx)
      m <- sapply(times, function(y) {
        sapply(1:ns, function(i) {
          w <- which(str.times[[i]] >= y)[1]
          ifelse(is.na(w), 0, str.n.risk[[i]][w])
        })
      })
      rownames(m) <- names(x$strata)
    } else {  # no strata
      m1 <- sapply(times, function(y) {
          w <- which(x$time >= y)[1]
          ifelse(is.na(w), 0, x$n.risk[w])
      })
      m <- matrix(m1, nrow = 1)
    }
    colnames(m) <- times
    m
}


addNrisk <- function(x, at = axTicks(1), 
      line = 4, hadj = 0.5, 
      title = 'Number at risk', title.adj = 0, 
      labels, hoff = 5, col = 1) {
    m <- nrisk(x, times = at)
    ns <- nrow(m)  # number of strata
    if(missing(labels)) {
      if(ns > 1) { 
        labels <- names(x$strata)
      } else {
        labels <- NA
      }
    }    
    label.pad <- paste(rep(' ', hoff), collapse = '') 
    labels2 <- paste(labels, label.pad, sep = '')
    labels2[is.na(labels)] <- NA
    col <- rep(col, length.out = ns)
    hasTitle <- (!is.null(title)) && (!is.na(title))
    if(hasTitle) {
      title(xlab = title, line = line, adj = title.adj)
    }
    for (i in 1:ns) {
      axis(1, at = at, labels = m[i,], 
        line = line + i + hasTitle - 2, tick = FALSE, 
        col.axis = col[i], hadj = hadj)
      axis(1, at = par('usr')[1], labels = labels2[i],
        line = line + i + hasTitle - 2, tick = FALSE,
        col.axis = col[i], hadj = 1)
    }
    invisible(m)
}



surviplot <- function(x, data = NULL, subset = NULL, 
     snames, stitle, 
     col, lty, lwd,
     show.nrisk = TRUE, color.nrisk = TRUE,
     hr.pos = 'topright', legend.pos = 'bottomleft', ...) {
  eval(bquote(s <- survfit(x, data = data, subset = .(substitute(subset)))))
  if('strata' %in% names(s)) {
    if(missing(stitle)) stitle <- strsplit(deparse(x), " ~ ")[[1]][2]
    if(missing(snames)) {
      snames <- names(s$strata)
      prefx <- paste(strsplit(deparse(x), " ~ ")[[1]][2], '=', sep = '')
      if(all(substr(snames, 1, nchar(prefx)) == prefx)) {
        snames <- substr(snames, nchar(prefx) + 1, 100)
      }
    }
    ns <- length(s$strata)
    stopifnot(length(snames) == ns)
  } else {    # no strata
    ns <- 1
    snames <- NA
    legend.pos <- NA
  }
  if(show.nrisk) {
    mar <- par('mar')
    mar[1] <- mar[1] + ns + 0.5
    opar <- par(mar = mar)
    on.exit(par(opar))
  }
  if(missing(col)) col <- 1:ns
  if(missing(lty)) lty <- 1
  if(missing(lwd)) lwd <- par('lwd')
  plot(s, col = col, lty = lty, lwd = lwd, ...)
  if(length(legend.pos) > 1 || !is.na(legend.pos)) {
    legend(legend.pos, legend = snames, title = stitle,
      col = col, lty = lty, lwd = lwd, bty = 'n')
  }
  if(ns == 2) {
    eval(bquote(cox <- summary(coxph(x, data = data, subset = .(substitute(subset))))))
    hr <- format(cox$conf.int[1, c(1, 3, 4)], digits = 2)
    p <- format(cox$sctest[3], digits = 2)
    txt1 <- paste('HR = ', hr[1], ' (', hr[2], ' - ', hr[3], ')', sep = '')
    txt2 <- paste('logrank P =', p)
    legend(hr.pos, legend = c(txt1, txt2), bty = 'n')
  }
  if(show.nrisk) {
    addNrisk(s, labels = snames,
      col = if(color.nrisk) col else 1)
  }
  if(ns == 2) return(invisible(c(txt1, txt2)))
}



censor <- function(x, at) {
  stopifnot(class(x) == 'Surv')
  stopifnot(attr(x, "type") == "right")
  wh <- x[, 'time'] > at
  x[wh, 'time'] <- at 
  x[wh, 'status'] <- 0
  x
}


cutn <- function(x, n = 2, ...) {
  stopifnot(n >= 1)
  nUnique <- length(na.omit(unique(x)))
  if(nUnique == 0) {
    return(factor(rep(NA, length(x))))
  } else {
    if(nUnique == 1) {
      breaks <- pretty(x)
    } else {
      breaks <-  quantile(x, probs = seq(0, 1, length = n + 1), na.rm = TRUE)
    }  
    return(cut(x, breaks = breaks, include.lowest = TRUE, ...))
  }
}


invert <- function(x) {
  stopifnot(class(x) == 'Surv')
  x[,2] <- 1 - x[,2]
  x
}






R/surviplot.R
# surviplot.R
#
# Aron Eklund   eklund@cbs.dtu.dk
#
# formerly known as "survplot.R" until 2016-08-08
#
#


nrisk <- function(x, times = pretty(x$time)) { 
    stopifnot(class(x) == 'survfit')
    if('strata' %in% names(x)) {
      ns <- length(x$strata)
      idx <- rep.int(1:ns, x$strata)
      str.n.risk <- split(x$n.risk, idx)
      str.times <- split(x$time, idx)
      m <- sapply(times, function(y) {
        sapply(1:ns, function(i) {
          w <- which(str.times[[i]] >= y)[1]
          ifelse(is.na(w), 0, str.n.risk[[i]][w])
        })
      })
      rownames(m) <- names(x$strata)
    } else {  # no strata
      m1 <- sapply(times, function(y) {
          w <- which(x$time >= y)[1]
          ifelse(is.na(w), 0, x$n.risk[w])
      })
      m <- matrix(m1, nrow = 1)
    }
    colnames(m) <- times
    m
}


addNrisk <- function(x, at = axTicks(1), 
      line = 4, hadj = 0.5, 
      title = 'Number at risk', title.adj = 0, 
      labels, hoff = 5, col = 1) {
    m <- nrisk(x, times = at)
    ns <- nrow(m)  # number of strata
    if(missing(labels)) {
      if(ns > 1) { 
        labels <- names(x$strata)
      } else {
        labels <- NA
      }
    }    
    label.pad <- paste(rep(' ', hoff), collapse = '') 
    labels2 <- paste(labels, label.pad, sep = '')
    labels2[is.na(labels)] <- NA
    col <- rep(col, length.out = ns)
    hasTitle <- (!is.null(title)) && (!is.na(title))
    if(hasTitle) {
      title(xlab = title, line = line, adj = title.adj)
    }
    for (i in 1:ns) {
      axis(1, at = at, labels = m[i,], 
        line = line + i + hasTitle - 2, tick = FALSE, 
        col.axis = col[i], hadj = hadj)
      axis(1, at = par('usr')[1], labels = labels2[i],
        line = line + i + hasTitle - 2, tick = FALSE,
        col.axis = col[i], hadj = 1)
    }
    invisible(m)
}



surviplot <- function(x, data = NULL, subset = NULL, 
     snames, stitle, 
     col, lty, lwd,
     show.nrisk = TRUE, color.nrisk = TRUE,
     hr.pos = 'topright', legend.pos = 'bottomleft', ...) {
  eval(bquote(s <- survfit(x, data = data, subset = .(substitute(subset)))))
  if('strata' %in% names(s)) {
    if(missing(stitle)) stitle <- strsplit(deparse(x), " ~ ")[[1]][2]
    if(missing(snames)) {
      snames <- names(s$strata)
      prefx <- paste(strsplit(deparse(x), " ~ ")[[1]][2], '=', sep = '')
      if(all(substr(snames, 1, nchar(prefx)) == prefx)) {
        snames <- substr(snames, nchar(prefx) + 1, 100)
      }
    }
    ns <- length(s$strata)
    stopifnot(length(snames) == ns)
  } else {    # no strata
    ns <- 1
    snames <- NA
    legend.pos <- NA
  }
  if(show.nrisk) {
    mar <- par('mar')
    mar[1] <- mar[1] + ns + 0.5
    opar <- par(mar = mar)
    on.exit(par(opar))
  }
  if(missing(col)) col <- 1:ns
  if(missing(lty)) lty <- 1
  if(missing(lwd)) lwd <- par('lwd')
  plot(s, col = col, lty = lty, lwd = lwd, ...)
  if(length(legend.pos) > 1 || !is.na(legend.pos)) {
    legend(legend.pos, legend = snames, title = stitle,
      col = col, lty = lty, lwd = lwd, bty = 'n')
  }
  if(ns == 2) {
    eval(bquote(cox <- summary(coxph(x, data = data, subset = .(substitute(subset))))))
    hr <- format(cox$conf.int[1, c(1, 3, 4)], digits = 2)
    p <- format(cox$sctest[3], digits = 2)
    txt1 <- paste('HR = ', hr[1], ' (', hr[2], ' - ', hr[3], ')', sep = '')
    txt2 <- paste('logrank P =', p)
    legend(hr.pos, legend = c(txt1, txt2), bty = 'n')
  }
  if(show.nrisk) {
    addNrisk(s, labels = snames,
      col = if(color.nrisk) col else 1)
  }
  if(ns == 2) return(invisible(c(txt1, txt2)))
}



censor <- function(x, at) {
  stopifnot(class(x) == 'Surv')
  stopifnot(attr(x, "type") == "right")
  wh <- x[, 'time'] > at
  x[wh, 'time'] <- at 
  x[wh, 'status'] <- 0
  x
}


cutn <- function(x, n = 2, right = TRUE, ...) {
    stopifnot(n >= 1)
    xUnique <- sort(na.omit(unique(x)))
    nUnique <- length(xUnique)
    if(n > nUnique) n <- nUnique
    if(nUnique == 0) {
        return(factor(rep(NA, length(x))))
    } else {
        if (nUnique == 1) {
            breaks <- pretty(x)
        } else if (nUnique == n) {
            if(right) {
                breaks <- c(pretty(extendrange(xUnique[1:2]))[1], xUnique)
            } else {
                breaks <- c(xUnique, tail(pretty(extendrange(tail(xUnique, n = 2))), n = 1))
            }
        } else {
            breaks <- quantile(x, probs = seq(0, 1, length = n + 1), na.rm = TRUE)
            ## try to fix some relatively common problems
            if(breaks[1] == breaks[2] && n >= 2) { 
                breaks[2] <- mean(c(breaks[2], min(x[x > breaks[2]], na.rm = TRUE)))
            }
            if (breaks[n] == breaks[n + 1] && n >= 2) { 
                breaks[n] <- mean(c(breaks[n], max(x[x < breaks[n]], na.rm = TRUE)))
            }
        }  
        return(cut(x, breaks = unique(breaks), include.lowest = TRUE, right = right, ...))
    }
}

invert <- function(x) {
  stopifnot(class(x) == 'Surv')
  x[,2] <- 1 - x[,2]
  x
}




